tests/validation/cattlevalidationtest/core/test_github.py

```diff
@@ -217,36 +220,121 @@ def test_github_add_whitelisted_user(github_client):
 
 @if_github
 def test_github_projects(github_client, cattle_url):
-    user_client = from_env(url=cattle_url)
-    switch_on_auth(github_client)
+    try:
+        user_client = from_env(url=cattle_url)
+        switch_on_auth(github_client)
 
-    #   set whitelisted orgs
-    github_client.create_githubconfig(allowedUsers=['ranchertest01'])
+        #   set whitelisted orgs
+        github_client.create_githubconfig(allowedUsers=['ranchertest01'])
 
-    #   test that these users were whitelisted
-    r = github_client.list_githubconfig()
+        #   test that these users were whitelisted
+        r = github_client.list_githubconfig()
 
-    users = r[0]['allowedUsers']
+        users = r[0]['allowedUsers']
 
-    assert 'ranchertest01' in users
+        assert 'ranchertest01' in users
 
-    rancherpass = os.getenv('API_AUTH_RANCHER_TEST_PASS', None)
+        rancherpass = os.getenv('API_AUTH_RANCHER_TEST_PASS', None)
 
-    if rancherpass is None:
-        assert False
+        if rancherpass is None:
+            assert False
 
-    new_token = github_request_code('ranchertest01', rancherpass)
-    new_token = github_request_token(new_token)
-    user_client._auth = GithubAuth(new_token)
-    projects = user_client.list_project()
+        new_token = github_request_code('ranchertest01', rancherpass)
+        new_token = github_request_token(new_token)
+        user_client._auth = GithubAuth(new_token)
+        projects = user_client.list_project()
+        try:
+            if len(projects) == 0:
+                uc = user_client
+                uc.create_project(externalIdType='project:github_user')
+                assert uc.externalId == 'ranchertest01'
+        except:
+            pass
+
+        projects = user_client.list_project()
+
+        assert len(projects) == 1
+        switch_off_auth(github_client)
+    except Exception as e:
+        switch_off_auth(github_client)
```

Comment:
This better accomplished with hooking into the py.test using finalizers http://pytest.org/latest/fixture.html#fixture-finalization-executing-teardown-code